students/perrymant/21-docker.md

```diff
@@ -0,0 +1,99 @@
+# Docker
+
+> Docker runs processes in isolated containers. A container is a process which runs on a host. The host may be local or remote. When an operator executes docker run, the container process that runs is isolated in that it has its own file system, its own networking, and its own isolated process tree separate from the host. - [Docker](https://docs.docker.com/)
+
+- What is a container? at its simplest it's a sandbox for a process - allowing the process to have its own namespace, restricting certain capabilities and imposing resource limits as required.
+- Container technology allows multiple isolated user space instances to be run on a single host.
+- Containers are generally considered a lightweight or lean technology because they require limited overhead, as each process can be run on very small Linux distributions.
+- Docker encourages microservices architecture.
+- Images are the building blocks of the Docker world. You launch your containers from images.
+- `image` vs `container`: an image is akin to an OOP class (the blueprint), and the container is akin to an OOP object (the instance).
+- Docker stores the images you build in registries. There are two types of registries: public and private, and can be found on (Docker Hub)[https://hub.docker.com], which contains over 10,000 images that other people have built and shared.
+- A Docker container comprises:
+    - An image format.
+    - A set of standard operations.
+    - An execution environment.
+
+### Overview of common commands
+
+```sh
+$ docker pull mongo
+$ docker run -d mongo
+$ docker ps
+$ docker rm (remove a container)
+$ docker images (ps, but for images)
+$ docker rmi (remove image)
+$ docker stats
+$ docker exec -it my_container_name mongo
+```
+
+### Commands: in depth
+- Firstly, the docker daemon needs to be running running, I use the `Docker Desktop` app for this.
+- Try `$ docker run hello-world` to run a very simple hello-world image.
+- `$ docker pull <name>`: To make sure you have the relevant files for the project, `docker pull <name>` will download a particular image, or set of images (i.e., a repository) from the Docker Hub registry. You can also use the docker `push` command, which pushes the image to its configured registry.
+- `$ docker run --name some-mongo -d mongo:tag`: Start a mongo server instance, where `some-mongo` is the name you want to assign to your container and tag is the tag specifying the MongoDB version you want. The `-d` option detaches the container, allowing it to run in the background.
+- `$ docker run -d MongoDB` will return the container ID.
```

Comment:
```suggestion
- `$ docker run -d MongoDB` will run container in background and return the container ID.
```
I think this is practically the same than the previous command, so you could delete this and put to the previous command that `docker run` returns the container ID.